### ISSUE TITLE
Model support: alpindale/dbrx-instruct

### ISSUE THREAD

**Issue by ftynse** [martin-luecke/rocm-systems#89]
- [x] unmodified model runs on gfx942 OR gfx950
- [ ] relevant kernels extracted (link to the hotswap-testing repository)
- [x] transpilation succeeds on gfx942 OR gfx950
- [x] individual kernels run with synthetic data
- [ ] transpiled kernels run and produces result within FP noise boundary on synthetic data (no NaNs)
- [x] whole model runs with transpiled kernels
- [x] relevant throughput metric (token/s, frame/s) measured and reported below (alongside git hashes used)
- [ ] all of the above on BOTH gfx942 AND gfx950

**Comment by ftynse** [comment]
https://huggingface.co/alpindale/dbrx-instruct

**Comment by tgymnich** [comment]
# DBRX Instruct SGLang Local vs HotSwap E2E Summary

- Root: `/tmp/sglang-dbrx-compare/sglang_dbrx_e2e_20260622_101709`
- Profile: `dbrx`
- GPU: `HIP_VISIBLE_DEVICES=0,1,2,3,4,5,6,7`
- Source target for HotSwap: `gfx1250:32`
- Target gfx: `gfx942`
- Max new tokens: `16`
- Prompt cases: `3`
- MoE runner backend: `None`
- Patch requirement: No GPT-OSS-specific patches required.
- CUDA/HIP graph capture: `disabled`

## Measured Warm Comparison
- Local generation tok/s: `5.754`
- HotSwap generation tok/s: `4.575`
- HotSwap / local generation ratio: `0.795`
- Local total tok/s including engine init: `0.505`
- HotSwap total tok/s including engine init: `0.502`

## Numerical Equivalence (local vs HotSwap)
- Verdict (warm/run2): `FAIL`
- Overall status (warm/run2): `diverged`
- Logprob tolerance (atol): `5.00e-03`
- Cases total: `3`
- Cases within tolerance: `0`
- Cases with token divergence: `2`
- Cases with argmax tie at first divergence: `1`
- Max chosen-token logprob diff (warm, full stream): `2.553`
- Max chosen-token logprob diff (warm, matching prefix): `0.252`
- Max top-K logprob diff (warm, full stream): `5.211`
- Max top-K logprob diff (warm, matching prefix): `0.748`
- KL tolerance: `1.00e-03`
- Cases distributionally equivalent only: `0`
- Teacher-forced divergence rate tolerance: `5.00e-02`
- Teacher-forced divergence count tolerance: `1`
- Cases teacher-forced equivalent only: `0`
- Cases diverged (post-reclassification): `2`
- Max forward KL (warm, comparability window): `0.454`
- Max reverse KL (warm, comparability window): `0.468`
- Max TVD (warm, comparability window): `0.461`

### Equivalence run1 (warmup)
- Overall: `diverged`
- Max chosen-token logprob diff (full stream / matching prefix): `2.553` / `0.252`
- Max top-K logprob diff (full stream / matching prefix): `5.211` / `0.748`
  - `identity_hotswap`: status=`diverged` first_div=`9` matched_prefix=`9`/`16` chosen_max=`2.553` chosen_max_prefix=`0.252` top_max=`2.864` top_max_prefix=`0.748` forward_kl_max=`0.454` reverse_kl_max=`0.468` tvd_max=`0.461` tie_at_first_div=true local_top1_top2_gap=`0.000e+00` divergent_tokens_shared_topk=`True`
  - `add_two_numbers`: status=`diverged` first_div=`5` matched_prefix=`5`/`16` chosen_max=`2.152` chosen_max_prefix=`0.159` top_max=`5.211` top_max_prefix=`0.235` forward_kl_max=`6.700e-03` reverse_kl_max=`6.630e-03` tvd_max=`0.054` divergent_tokens_shared_topk=`True`
  - `subtract_two_numbers`: status=`numerically_close` first_div=`None` matched_prefix=`16`/`16` chosen_max=`0.176` chosen_max_prefix=`0.176` top_max=`0.574` top_max_prefix=`0.574` forward_kl_max=`0.029` reverse_kl_max=`0.028` tvd_max=`0.111`

### Equivalence run2 (measured warm)
- Overall: `diverged`
- Max chosen-token logprob diff (full stream / matching prefix): `2.553` / `0.252`
- Max top-K logprob diff (full stream / matching prefix): `5.211` / `0.748`
  - `identity_hotswap`: status=`diverged` first_div=`9` matched_prefix=`9`/`16` chosen_max=`2.553` chosen_max_prefix=`0.252` top_max=`2.864` top_max_prefix=`0.748` forward_kl_max=`0.454` reverse_kl_max=`0.468` tvd_max=`0.461` tie_at_first_div=true local_top1_top2_gap=`0.000e+00` divergent_tokens_shared_topk=`True`
  - `add_two_numbers`: status=`diverged` first_div=`5` matched_prefix=`5`/`16` chosen_max=`2.152` chosen_max_prefix=`0.159` top_max=`5.211` top_max_prefix=`0.235` forward_kl_max=`6.700e-03` reverse_kl_max=`6.630e-03` tvd_max=`0.054` divergent_tokens_shared_topk=`True`
  - `subtract_two_numbers`: status=`numerically_close` first_div=`None` matched_prefix=`16`/`16` chosen_max=`0.176` chosen_max_prefix=`0.176` top_max=`0.574` top_max_prefix=`0.574` forward_kl_max=`0.029` reverse_kl_max=`0.028` tvd_max=`0.111`

## Teacher-Forced Divergence Check
- Status: `completed`
- Cases total: `3`
- Cases with direct divergence checked: `2`
- Forced prefixes checked: `48`
- Forced next-token divergences: `16`
- Max forced-next logprob diff: `0.470`
- Max common top-K logprob diff: `0.969`
- First forced divergence: case=`add_two_numbers` trajectory=`local` step=`1` local=`949`/`None` hotswap=`0`/`None`
  - `add_two_numbers`: prefixes=`26` forced_divergences=`9` forced_lp_max=`0.201` top_max=`0.840` reclassified=`false` rate=`0.346` (tol=`0.050`) count=`9` (tol=`1`)
  - `identity_hotswap`: prefixes=`22` forced_divergences=`7` forced_lp_max=`0.470` top_max=`0.969` reclassified=`false` rate=`0.318` (tol=`0.050`) count=`7` (tol=`1`)

## local
### run1 (warmup)
- Generation tok/s: `4.989`
- Total tok/s including engine init: `0.503`
- generate_s: `9.623s`
- engine_init_s: `82.139s`
- total_s: `95.439s`

### run2 (measured warm)
- Generation tok/s: `5.754`
- Total tok/s including engine init: `0.505`
- generate_s: `8.345s`
- engine_init_s: `83.078s`
- total_s: `95.115s`


## hotswap
### run1 (warmup)
- Generation tok/s: `2.806`
- Total tok/s including engine init: `0.464`
- generate_s: `17.110s`
- engine_init_s: `82.850s`
- total_s: `103.406s`
- Proof status: `translated`
- HotSwap proof: `120/120` success
- Triton target override events: `19`
- HotSwap target gfx: `['gfx942']`
- HotSwap cache hits/misses: `57/63`

### run2 (measured warm)
- Generation tok/s: `4.575`
- Total tok/s including engine init: `0.502`
- generate_s: `10.495s`
- engine_init_s: `81.641s`
- total_s: `95.668s`
- Proof status: `translated`
- HotSwap proof: `120/120` success
- Triton target override events: `19`
- HotSwap target gfx: `['gfx942']`
- HotSwap cache hits/misses: `120/0`

**Comment by suphilip** [comment]
## dbrx-instruct — HotSwap e2e close (both arches)

`alpindale/dbrx-instruct` (DbrxForCausalLM, Databricks MoE ~132B) runs coherent, stable, and fault-free as a **whole model** through HotSwap on **both gfx950 and gfx942** from a single `gfx1250:32` source. Every leg is gated on a native oracle (refuse-don't-miscompile held; no forced pass).

### Summary — 8 model-support criteria

| # | Criterion | gfx950 (MI350X) | gfx942 (MI300X) |
|---|-----------|---|---|
| 1 | Unmodified model runs natively | ✅ | ✅ |
| 2 | Kernels extracted (count > 0) | ✅ 16 distinct / 186 events | ✅ 13 distinct |
| 3 | Transpile succeeds (0 refused) | ✅ 186 ok / 0 fail / 0 refused | ✅ 24/24, 32/32, 0 refused |
| 4 | Transpiled kernels run (0 fault) | ✅ 0 HIP-719 / 0 fault (scaled dispatch) | ✅ 0 HIP-719 / 0 fault |
| 5 | Transpiled == native (FP-noise, no NaN) | ✅ argmax + top-10 set = native, stable 3/3 | ✅ argmax + top-10 set = native |
| 6 | Whole model runs through HotSwap | ✅ coherent + stable | ✅ coherent + stable |
| 7 | Throughput (overhead vs native, not raw serving) | ✅ ~0% overhead: on-GPU native ~2.07 vs HotSwap ~2.25 tok/s ¹ | ✅ no transpiler-attributable overhead ² |
| 8 | Both gfx942 AND gfx950 | ✅ | ✅ |

¹ Earlier scaled leg, 8 new tokens, warmup-inclusive; the option-B close leg itself recorded only model load (83.4s).
² e2e decode is eager-attention + CPU-offload bound — HotSwap ~0.06 vs native ~0.084 tok/s reflects offload latency (dbrx ~245 GB, 15/43 layers on CPU on the 192 GiB MI300X), not transpiler overhead.

The two arches pass by **different translation paths**, because the wave32→wave64 reduction behaves differently on each:

- **gfx950 — scaled dispatch.** Its phantom-lane RMSNorm reduction miscompiles as WaveNative, so kernels are scaled (one source wave per target wave). This requires the inductor toggle `TORCHINDUCTOR_USE_STATIC_CUDA_LAUNCHER=0` so the scaled launch is actually consumed, otherwise HIP-719 (see Known limitations).
- **gfx942 — WaveNative (no scaling).** WaveNative is already bit-coherent here, and force-scaling it miscompiles, so no scaling is applied.

On both paths the translated model matches the native oracle: greedy argmax **264** (France) / **220** (arithmetic) = native, top-10 logit index set = native, coherent English (gfx950 stable across 3 repeats). The bar is coherence, not bit-identity: dbrx's MoE graph is run-to-run non-deterministic even natively (`torch.where`/nonzero graph-break + inductor autotune + rocBLAS). Throughput (row 7) is reported as **HotSwap overhead vs the native oracle**, not raw serving tok/s: only fused pointwise/reduction ops are transpiled (matmuls stay on rocBLAS/ATen), so the e2e decode figures are dominated by eager attention and, on gfx942, CPU offload. Cite as "overhead = native".

### Environment (both runs)

`torch=2.11.0+gitd0c8b1f`, `hip=7.2.53211` (ROCm 7.2.3), `transformers==4.44.2`, dtype bf16, eager attention, `torch.compile(model.forward)`, greedy, `trust_remote_code=1`.

### Provenance (md5 = binary proof; SHAs identify source trees only)

**gfx950 (scaled dispatch, MI350X):**
```
COMGR  libamd_comgr_hotswap.so.3  md5 e53cc27ab1b91bab316619f4399acfdb
       (martin-luecke/llvm-project users/ftynse/hotswap-modrep-doubled-dispatch,
        #283-HEAD tip 23b77403e + env force-scaled trigger)
ROCr   libhsa-runtime64.so.1      md5 ca2e23a99fcf0ab27738e7fcb51cb64c (abi723 scaled AQL consumer)
intercept libhotswap_intercept.so md5 6b3e8be51c2b660857d519b013acab8b
```

**gfx942 (WaveNative, MI300X):**
```
ROCr   libhsa-runtime64.so.1      md5 ca2e23a99fcf0ab27738e7fcb51cb64c
COMGR  libamd_comgr_hotswap.so.3  md5 947c73044d44510c089d548afcd6233e
intercept libhotswap_intercept.so md5 756c31d080b18fc799e02b2b84364f1a
raise_cli                         md5 8c1dc206f075cf0d412f70f964dbcc6a
```

ROCr source: martin-luecke/rocm-systems `users/mluecke/hotswap-compatibility@b38dca55de` + the AQL scaled-dispatch consumer / kernel-identity bridge. The git SHAs come from the build record plus the current checkout state; the md5s are the direct binary proof — for bit-exact provenance, trust the md5s.

### Known limitations / caveats (carried forward)

- **gfx950 env toggle.** The close is gated on `TORCHINDUCTOR_USE_STATIC_CUDA_LAUNCHER=0` (a supported inductor env toggle, not a code hack), required only for the scaled wave32→wave64 reduction class: it routes launches off torch's compiled static launcher and through the HSA queue the ROCr AQL scaled consumer wraps, so reductions launch at wg=512 and the HIP-719 gate is cleared. Durable follow-up: teach `_StaticCudaLauncher` to honor the transpiled kernel's declared `max_flat_workgroup_size` (512) so the fast static path also scales. gfx942 needs none of this.
- **Footprint / offload.** dbrx bf16 ~245 GB — fits one GPU only on the 288 GiB MI350X; on MI300X (192 GiB) and 252 GiB MI350X it needs `device_map=auto` CPU offload, so tok/s is dominated by offload latency, not HotSwap.
- **Coverage by construction.** Matmuls dispatch to rocBLAS/ATen; only the fused pointwise/reduction ops become gfx1250 Triton kernels that HotSwap translates. CPU-offloaded layers run natively on CPU and are not HotSwap-eligible.
- **transformers pin.** `transformers==4.44.2` is required; 5.x and 4.57.1 both fail on dbrx's 4.38-vintage modeling (Cache-API cascade).

### Status

E2E-validated coherent + stable on both arches; refuse-don't-miscompile held (native oracle gated every leg).
